man1/gnubiff.1
.TH gnubiff 1 " 8 September 2007" "Version 2.3.0" "gnubiff Manual Pages"
.SH NAME
gnubiff \- a mail notification program supporting GTK+ and GNOME
.SH SYNOPSIS
.B gnubiff
.RI [ options ]
.SH DESCRIPTION
gnubiff is a mail notification program that checks for mail, displays
headers when new mail has arrived and allow to read first lines of new
messages. It relies on the GNOME and GTK libraries but can be compiled
and used with or without GNOME support. Supported protocols are pop3,
apop, imap4, mh, qmail and mailfile.  Furthermore, gnubiff is fully
configurable with a lot of options like polltime, poptime, sounds, mail
reader, mailbox names, etc. and can also filter spam. Note that
gnubiff is not a mail reader even if it offers the possibility of
reading very first lines of new messages. If you send some requests for
specific features to be implemented, keep that in mind!
.SH OPTIONS
gnubiff accepts the following command line options:
.TP
.BI "\-c " filename, " \-\-config " filename
Use the specified file as configuration file (default is ~/.gnubiffrc)
.TP
.B \-n, \-\-noconfigure
Skip the configure process
.TP
.B \-\-nogui
Start gnubiff without GUI
.TP
.B \-\-systemtray
Start gnubiff in the system tray
.TP
.B \-?, \-\-help
Show summary of options
.TP
.B  \-v, \-\-version
Print version information
.SH FILES
.TP
.I ~/.gnubiffrc
user specific options file
.SH "SEE ALSO"
More documentation about gnubiff can be found in the info pages with the command
.B "info gnubiff"
.SH BUGS
None known.
.SH AUTHORS
gnubiff was developed by Nicolas Rougier
<Nicolas.Rougier@loria.fr> and is currently maintained by Robert Sowada
<sowadart@gmx.de>. This manpage was written by Roland Stigge
<stigge@antcom.de> for Debian. It can be used elsewhere, of course.


NEWS
gnubiff -- History of visible changes.

Copyright (C) 2000-2007 Nicolas Rougier <Nicolas.Rougier@loria.fr>,
              2004-2007 Robert Sowada <sowadart@gmx.de>
See the end for copying conditions.
Please send gnubiff bug reports to gnubiff-bugs@lists.sourceforge.net

Version 2.2.8
This is a minor feature release. Changes:
 * Upgraded license to GPL version 3
 * Updated Swedish translation (thanks to Daniel Nylander)
 * Updated Vietnamese translation (thanks to Clytie Siddall)
 * Updated German translation

Version 2.2.7
This is a minor feature release. Changes:
 * Showing/Hiding popup via SIGUSR signals
 * Options for showing/hiding popup/applet in the taskbar
 * Several small bugfixes

Version 2.2.6
This is a minor feature release. Changes:
 * Updated Swedish translation (thanks to Daniel Nylander)
 * A small bugfix

Version 2.2.5
This is a minor bugfix release. Changes:
 * Fixed some bugs

Version 2.2.4
This is a minor bugfix release. Changes:
 * Some bug fixes
 * Portuguese translation now available (thanks to Gunther Furtado)

Version 2.2.3
This is a minor bugfix release. Changes:
 * Passwords longer than 15 characters are now handled properly
 * Compiling without fam.h is now possible
 * Some other bug fixes

Version 2.2.2
This is a minor feature release. Changes:
 * Some bug fixes
 * Updated Hungarian translation (thanks to Wallner Adam)
 * Updated Vietnamese translation (thanks to Clytie Siddall)
 * Options for filtering messages via regular expressions

Version 2.2.1
This is a minor feature release. Changes:
 * Some bug fixes
 * It is now possible to compile gnubiff without support for FAM
 * New option for not showing the preferences dialog at startup as default
 * Updated Vietnamese translation (thanks to Clytie Siddall)

Version 2.2.0
This is a major feature release. Changes:
 * Lots of bugfixes and code cleanups
 * gnubiff applet can be put in the system tray
 * gnubiff can be run without GUI or X
 * USR1 and USR2 signals are now monitored
 * Saved passwords are now encrypted via AES
 * Decoding of base64 encoded message bodies
 * About dialog for Gnome version

Version 2.1.9
This is a minor bugfix release. Changes:
 * Better detection of FAM/GAMIN at configuration time
 * FAM can be disabled for monitoring mailboxes (and polling be used instead)
 * Some additional FAM bugfixes
 * Some other small bugfixes

Version 2.1.8
This is a minor bugfix release. Changes:
 * Updated Italian translation (thanks to Stefano Fabri)
 * Added Hungarian translation (thanks to Wallner Adam)
 * Some small bugfixes

Version 2.1.7
This is a minor bugfix release. Changes:
 * Fixed a bug that prevented 2.1.6 to be installed on certain systems

Version 2.1.6
This is a minor bugfix release. Changes:
 * The gnome version of gnubiff can now be installed to any directory
    by any user

Version 2.1.5
This is a minor bugfix release. Changes:
 * Fixed some FAM and GAMIN related problems
 * Fixed problems when displaying non ASCII characters when using non UTF-8
   locales
 * Some other bugfixes

Version 2.1.4
This is a major bugfix release. Changes:
 * Support for mh mailboxes as used by Sylpheed
 * Handling of mh protocols optimized
 * Czech translation (thanks to Josef Vybiral)
 * Compliance with RFC 2047
 * Race conditions in FAM code fixed
 * Various other bugfixes

Version 2.1.3
This is a major bugfix release. Changes:
 * Maildir protocol fixed (thanks to Alexis S. L. Carvalho)
 * Mh protocol fixed
 * Fixes for 64 bit systems (thanks to David Mosberger)
 * Various other bugfixes
Note: The default behavior when using file protocol has changed. The access
time of the mail file is no longer reset by default. This can be changed by
setting the option "file_restore_atime" to true.

Version 2.1.2
This is a major bugfix release. Changes:
 * Many bug fixes
 * Improved displaying of messages
 * Vietnamese translation available

Version 2.1.1
Major bugfix release. Many thanks to Byron Foster and Calum Mackay for their
help in creating this release. Changes:
 * Many bug fixes
 * Review of network code
 * Applet and Popup window can be set to be sticky, kept above all other
   windows and not appear in pagers
 * Idling can be disabled (IMAP4)
 * Mails are now only fetched once per session
 * Out of the box support for FreeBSD (thanks to Tim Bishop)
 * Expert editing mode for all options provided by gnubiff (including internal
   options)
 * Sorting order of headers in popup can be configured

Version 2.1.0
 * Nicolas Rougier
   - New & simpler interface
   - Support of the IDLE state for imap4 (no more polling)
   - Support of the FAM (File Alteration Monitor)
     (no more polling for file/folder protocols)
   - Better detection of mailbox format
   - Bugfix: IMAP4 idle state test to avoid checking twice
   - Bugfix: Popup was shown on exiting preferences
   - Put back credits
   - Updated French translations
   - Fix a bug with IMAP4 idle feature (not recognizing the BYE command)
   - Added a TODO file to keep track of features/bugs to be implemented/fixed
   - Bugfix: Testing of "\"text\" \"plain\"" in imap4 parsing was wrong
             (some servers answer using upper cases letter)
   - Modified credits apperance in the preferences panel (and remove the
     about menu from gnome applet popup menu)
   - Bugfix: socket was closed twice when lock was lost in IDLE state
   - Bugfix: modify the configure script to have errors if popt or fam
             libraries not present
   - Sorted the contributors list
   - Feature: When adding a mailbox and if one is selected, the new mailbox is
	      copied from the selected one
   - Feature: If password is empty, gnubiff try to find by looking at
              other mailboxes and matching (address/username)
   - Updated THANKS file
   - Bugfix: display last received mail for each mailbox (up to	popup_size)
   - Bugfix: Take popup_use_size_ into account
   - Bugfix: Use of the decoded charset for imap4

src/gnubiff.cc
// ========================================================================
// gnubiff -- a mail notification program
// Copyright (c) 2000-2007 Nicolas Rougier, 2004-2007 Robert Sowada
//
// This program is free software: you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
// ========================================================================
//
// File          : $RCSfile$
// Revision      : $Revision$
// Revision date : $Date$
// Author(s)     : Nicolas Rougier, Robert Sowada
// Short         : 
//
// This file is part of gnubiff.
//
// -*- mode:C++; tab-width:4; c-basic-offset:4; indent-tabs-mode:nil -*-
// ========================================================================

#include "support.h"

#include <gtk/gtk.h>
#include <popt.h>

#ifdef USE_GNOME
#	include <gnome.h>
#	include <panel-applet.h>
#	include "ui-applet-gnome.h"
#endif

#include "biff.h"
#include "nls.h"
#include "ui-preferences.h"
#include "ui-applet.h"
#include "ui-applet-gtk.h"

int main (int argc, char **argv);
int mainGNOME (int argc, char **argv);


int main (int argc, char **argv) {
	// Initialize i18n support
	setlocale (LC_ALL, "");
#ifdef ENABLE_NLS
	bindtextdomain (GETTEXT_PACKAGE, GNUBIFF_LOCALEDIR);
#   ifdef HAVE_BIND_TEXTDOMAIN_CODESET
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
#   endif
	textdomain (GETTEXT_PACKAGE);
#endif

	// Thread initialization
	g_thread_init (NULL);
	gdk_threads_init ();

#ifdef USE_GNOME
	// Test for presence of "--oaf-ior-fd" option to determine if gnubiff is
	// started in gnome panel mode
 	for (gint i=0; i<argc; i++)
		if (std::string(argv[i]).find("--oaf-ior-fd")==0)
			return mainGNOME (argc, argv);
#endif

	//
	// Parse Options
	//
	GMainLoop*  gmainloop = NULL;
	poptContext poptcon;
	guint ui_mode = MODE_GTK;
	int status;
	char *config_file = 0;
	int no_configure = false, print_version = false, no_gui = false;
	int systemtray = false;
#if defined DEBUG && defined USE_GNOME
	int debug_applet=false;

   	static struct poptOption options_debug[] =
	{
	   	{"applet",'\0', POPT_ARG_NONE,   &debug_applet,  0,
		 N_("Start gnome applet from command line"), NULL},
		POPT_TABLEEND
	};
#endif
	static struct poptOption options_general[] =
	{
	   	{"config",      'c' , POPT_ARG_STRING, &config_file,   0,
		 N_("Configuration file to use"),  N_("file")},
		{"noconfigure", 'n' , POPT_ARG_NONE,   &no_configure,  0,
		 N_("Skip the configuration process"), NULL},
		{"nogui",		'\0', POPT_ARG_NONE,   &no_gui,		   0,
		 N_("Start gnubiff without GUI"), NULL},
		{"systemtray",	'\0', POPT_ARG_NONE,   &systemtray,	   0,
		 N_("Put gnubiff's icon into the system tray"), NULL},
		{"version",     'v' , POPT_ARG_NONE,   &print_version, 0,
		 N_("Print version information and exit"), NULL},
		POPT_TABLEEND
	};
	static struct poptOption options[] =
	{
	   	{NULL, '\0', POPT_ARG_INCLUDE_TABLE, &options_general, 0,
		 N_("General command line options:"), NULL },
#if defined DEBUG && defined USE_GNOME
		{NULL, '\0', POPT_ARG_INCLUDE_TABLE, &options_debug, 0,
		 N_("Options for debugging:"), NULL },
#endif
		POPT_AUTOHELP
		POPT_TABLEEND
	};

	// Parse command line
	poptcon = poptGetContext ("gnubiff", argc,  (const char **) argv,
							  options, 0);
	while ((status = poptGetNextOpt(poptcon)) >= 0);
	if (status < -1) {
		fprintf (stderr, "%s: %s\n\n",
				 poptBadOption (poptcon, POPT_BADOPTION_NOALIAS),
				 poptStrerror (status));
		poptPrintHelp (poptcon, stderr, 0);
		exit (1);
	}
	poptGetNextOpt(poptcon);

	// Decide which frontend to use
	if (no_gui) {
		ui_mode = MODE_NOGUI;
		no_configure = true;
	}
	if (systemtray)
		ui_mode = MODE_SYSTEMTRAY;

#if defined DEBUG && defined USE_GNOME
	if (debug_applet)
		return mainGNOME (argc, argv);
#endif
	
	// Initialization of Glib and (if needed) GTK
	if (no_gui)
		gmainloop = g_main_loop_new (NULL, true);
	else
		gtk_init (&argc, &argv);

	// Print version information if requested and exit
	if (print_version) {
#ifdef IS_CVS_VERSION
		g_print ("gnubiff version "PACKAGE_VERSION" CVS\n");
#else
		g_print ("gnubiff version "PACKAGE_VERSION"\n");
#endif
		exit (EXIT_SUCCESS);
	}

	// Create biff with configuration file (or not)
	Biff *biff;
	if (config_file)
		biff = new Biff (ui_mode, config_file);
	else
		biff = new Biff (ui_mode);

	// Decide whether the preferences dialog shall be shown now
	no_configure = no_configure || !biff->value_bool ("startup_preferences");

	// Start applet
	biff->applet()->start (!no_configure);

	// Main loop
	if (no_gui)
		g_main_loop_run (gmainloop);
	else {
		gdk_threads_enter();
		gtk_main();
		gdk_threads_leave();
	}

	// Exit
	return 0;
}

#ifdef USE_GNOME

int mainGNOME (int argc, char **argv) {
#if defined(PREFIX) && defined(SYSCONFDIR) && defined(DATADIR) && defined(LIBDIR)
	gnome_program_init ("gnubiff", "0", LIBGNOMEUI_MODULE, argc, argv,
						GNOME_PROGRAM_STANDARD_PROPERTIES, NULL);
#else
	gnome_program_init ("gnubiff", "0", LIBGNOMEUI_MODULE, argc, argv,
						GNOME_PARAM_NONE);
#endif

	panel_applet_factory_main ("OAFIID:GNOME_gnubiffApplet_Factory",
							   PANEL_TYPE_APPLET,
							   AppletGnome::gnubiff_applet_factory, 0);

	return 0;
}
#endif
